zcash_client_backend/src/keys.rs

```diff
@@ -144,16 +195,151 @@ impl UnifiedSpendingKey {
     pub fn orchard(&self) -> &orchard::keys::SpendingKey {
         &self.orchard
     }
+
+    /// Returns a binary encoding of this key suitable for decoding with [`decode`].
+    ///
+    /// The encoded form of a unified spending key is only intended for use
+    /// within wallets when required for storage and/or crossing FFI boundaries;
+    /// unified spending keys should not be exposed to users, and consequently
+    /// no string-based encoding is defined. This encoding does not include any
+    /// internal validation metadata (such as checksums) as keys decoded from
+    /// this form will necessarily be validated when the attempt is made to
+    /// spend a note that they have authority for.
+    #[cfg(feature = "unstable")]
+    pub fn to_bytes(&self, epoch: Epoch) -> Vec<u8> {
+        let mut result = vec![];
+        result.write_u32::<LittleEndian>(epoch.id()).unwrap();
+
+        // orchard
+        let orchard_key = self.orchard();
+        CompactSize::write(&mut result, usize::from(Typecode::Orchard)).unwrap();
+
+        let orchard_key_bytes = orchard_key.to_bytes();
+        CompactSize::write(&mut result, orchard_key_bytes.len()).unwrap();
+        result.write_all(orchard_key_bytes).unwrap();
+
+        // sapling
+        let sapling_key = self.sapling();
+        CompactSize::write(&mut result, usize::from(Typecode::Sapling)).unwrap();
+
+        let sapling_key_bytes = sapling_key.to_bytes();
+        CompactSize::write(&mut result, sapling_key_bytes.len()).unwrap();
+        result.write_all(&sapling_key_bytes).unwrap();
+
+        // transparent
+        #[cfg(feature = "transparent-inputs")]
+        {
+            let account_tkey = self.transparent();
+            CompactSize::write(&mut result, usize::from(Typecode::P2pkh)).unwrap();
+
+            let account_tkey_bytes = account_tkey.to_bytes();
+            CompactSize::write(&mut result, account_tkey_bytes.len()).unwrap();
+            result.write_all(&account_tkey_bytes).unwrap();
+        }
+
+        result
+    }
+
+    #[allow(clippy::unnecessary_unwrap)]
+    #[cfg(feature = "unstable")]
+    pub fn from_bytes(epoch: Epoch, encoded: &[u8]) -> Result<Self, DecodingError> {
+        let mut source = std::io::Cursor::new(encoded);
+        let decoded_epoch = source
+            .read_u32::<LittleEndian>()
+            .map_err(|_| DecodingError::ReadError("epoch"))
+            .and_then(|id| Epoch::try_from_id(id).ok_or(DecodingError::EpochInvalid))?;
+
+        if decoded_epoch != epoch {
+            return Err(DecodingError::EpochMismatch(decoded_epoch));
+        }
+
+        let mut orchard = None;
+        let mut sapling = None;
+        #[cfg(feature = "transparent-inputs")]
+        let mut transparent = None;
+        loop {
```

Comment:
This loop allows the components to be read in any order. I guess that's fine, since we don't really care that this encoding is visibly canonical, since it is not visible to users.

Reply:
Yes, this is intentional.